plugins/flatpak/gs-flatpak.c

```diff
@@ -2567,34 +2567,31 @@ gs_flatpak_refine_runtime_for_install (GsFlatpak *self,
 				       GError **error)
 {
 	GsApp *runtime;
+	gsize len;
+	const gchar *str;
+	g_autoptr(GBytes) data = NULL;
 
-	/* only required for ostree-based flatpak apps */
-	if (gs_flatpak_app_get_file_kind (app) == GS_FLATPAK_APP_FILE_KIND_BUNDLE) {
-		gsize len;
-		const gchar *str;
-		g_autoptr(GBytes) data = NULL;
-		g_autoptr(GsApp) runtime_new = NULL;
-
-		data = gs_flatpak_fetch_remote_metadata (self, app, cancellable, error);
-		if (data == NULL) {
-			gs_utils_error_add_unique_id (error, app);
-			return FALSE;
-		}
+	if (gs_app_get_kind (app) != AS_APP_KIND_DESKTOP)
+		return TRUE;
 
-		str = g_bytes_get_data (data, &len);
-		runtime_new = gs_flatpak_create_runtime_from_metadata (self, app,
-								       str, len,
-								       error);
-		if (runtime_new == NULL)
-			return FALSE;
-		gs_app_set_update_runtime (app, runtime_new);
+	/* ensure that we get the right runtime that will need to be installed */
+	data = gs_flatpak_fetch_remote_metadata (self, app, cancellable, error);
+	if (data == NULL) {
+		gs_utils_error_add_unique_id (error, app);
+		return FALSE;
 	}
 
+	str = g_bytes_get_data (data, &len);
+	runtime = gs_flatpak_create_runtime_from_metadata (self, app,
+							   str, len,
+							   error);
+
 	/* no runtime required */
-	runtime = gs_app_get_update_runtime (app);
 	if (runtime == NULL)
 		return TRUE;
```

Comment:
Shouldn't this return `FALSE` now, as in https://github.com/endlessm/gnome-software/pull/274/commits/d7073b4e11d47e0f15da13076fbddc0fa1110862#diff-e64404ac7c111b350b85abaa4fb8084dL2589 ? It seems odd to me to return `TRUE` here if the call to `gs_flatpak_create_runtime_from_metadata()` failed (seemed fine with the old call to `gs_app_get_update_runtime()`, though)

Reply:
Nope. The value is right because in certain functions in GNOME Software's plugins, the `TRUE` value is also for when we want to ignore/skip. As you can see, we are also returning `TRUE` in:
```
if (gs_app_get_kind (app) != AS_APP_KIND_DESKTOP)
		return TRUE;
```
So in this case, just like before, if the app in question has no runtime, we return `TRUE`.
I know this is counterintuitive but I prefer if we go with this, as a proper fix (let libflatpak care about installing the runtime itself) will anyhow involve new API that we're planning and thus IMO we shouldn't make deep changes in how g-s is working ATM.

Reply:
> So in this case, just like before, if the app in question has no runtime, we return TRUE.
I know this is counterintuitive but I prefer if we go with this, as a proper fix (let libflatpak care about installing the runtime itself) will anyhow involve new API that we're planning and thus IMO we shouldn't make deep changes in how g-s is working ATM.

Yes, I get that. But in this case what I'm seeing is that you're returning TRUE when you **failed to create** a runtime for the app, not when a call to `gs_app_get_update_runtime()` has failed because there's no associated runtime, that's what I thought returning FALSE (as in other cases of failure before that point) made sense.

That said, I'm not extremely familiar with this code, so you probably know better and if you still think it's correct, I guess we can merge it. But first, can you confirm that?

Reply:
Ah, okay, I see it now. We should definitely do more than returning FALSE, I will need to check the implications of returning FALSE though. I will take a look. Thanks.